src/structstore_shared.hpp

```diff
@@ -163,72 +191,76 @@ class StructStoreShared {
             throw std::runtime_error("original size is invalid");
         }
 
-        lseek(shm_fd, 0, SEEK_SET);
-        shm_ptr = (SharedData*) mmap(
+        lseek(fd, 0, SEEK_SET);
+        ptr = (SharedData*) mmap(
                 original_ptr,
                 size,
                 PROT_READ | PROT_WRITE,
                 // ensures that the shared memory segment is mapped
                 // to the same region of memory for all processes
                 // the memory allocator relies on that
                 MAP_SHARED | MAP_FIXED_NOREPLACE,
-                shm_fd,
+                fd,
                 0);
 
-        if (shm_ptr == MAP_FAILED || shm_ptr != original_ptr) {
+        if (ptr == MAP_FAILED || ptr != original_ptr) {
             throw std::runtime_error("mmap'ing existing memory failed");
         }
 
-        if (shm_ptr->original_ptr != original_ptr) {
+        if (ptr->original_ptr != original_ptr) {
             throw std::runtime_error("inconsistency detected");
         }
 
-        ++shm_ptr->usage_count;
+        ++ptr->usage_count;
     }
 
 public:
 
     bool valid () {
 
-        return !shm_ptr->invalidated.load();
+        return !ptr->invalidated.load();
     }
 
     bool revalidate (bool block = true) {
 
-        if (!shm_ptr->invalidated.load()) {
+        if (!ptr->invalidated.load()) {
             return true;
         }
 
         // need to revalidate the shared memory segment
 
-        int new_shm_fd = -1;
+        int new_fd = -1;
 
         do {
-            if (new_shm_fd == -1) {
-                new_shm_fd = shm_open(shm_path.c_str(), O_RDWR, 0600);
-                if (new_shm_fd == -1) {
+            if (new_fd == -1) {
+                if (use_file) {
+                    new_fd = open(path.c_str(), O_RDWR, 0600);
+                } else {
+                    new_fd = shm_open(path.c_str(), O_RDWR, 0600);
+                }
+                if (new_fd == -1) {
                     continue;
                 }
             }
 
-            struct stat shm_stat = {};
-            fstat(shm_fd, &shm_stat);
+            struct stat fd_stat = {};
+            fstat(fd, &fd_stat);
```

Comment:
Just noticed this, I guess this should `fstat()` on `new_fd`? (this is an issue on current develop as well)

Reply:
Absolutely correct! That could have turned out nasty. Should be fixed now.